book/src/dev/rfcs/0012-value-pools.md

```diff
@@ -0,0 +1,437 @@
+        Transaction::V4 { .. } => {},
+        Transaction::V5 { .. } => {
+            
+        },
+    }
+}
+```
+
+### Create a method in `Block` that returns `ValueBalance<NegativeAllowed>` for the block
+
+- Method location is at `zebra-chain/src/block.rs`.
+- Method will make use of `Transaction::value_balance` method created before.
+
+```rust
+/// utxos must contain the utxos of every input in the block
+/// TODO: what about UTXOs created and spent in the block?
+pub fn value_balance(&self, utxos: &HashMap<transparent::OutPoint, Utxo>) -> ValueBalance<NegativeAllowed> {
+    self.transactions()
+        .map(Transaction::value_balance)
+        .sum()
+        .unwrap_or_else(ValueBalance::default)
+}
+```
+
+### Pass the value balance for this block from the consensus into the state
+
+- Add a new field into `PreparedBlock` located at `zebra-state/src/request.rs`, this is the `NonFinalized` section of the state.
+
+```rust
+pub struct PreparedBlock {
+    ..
+    /// The value balances for each pool for this block.
+    pub block_value_balance: ValuePool<NegativeAllowed>,
+}
+```
+- In `zebra-consensus/src/block.rs` pass the value balance to the zebra-state:
+
+```rust
+let block_value_balance = block.value_balance();
+let prepared_block = zs::PreparedBlock {
+    ..
+    block_value_balance,
+};
+```
+
+### Add a value pool into the state `Chain` struct
+
+- This is the value pool for the non finalized part of the blockchain.
+- Location of the `Chain` structure where the pool field will be added: `zebra-state/src/service/non_finalized_state/chain.rs`
+
+```rust
+pub struct Chain {
+    ..
+    value_pool: ValueBalance<NonNegative>,
+}
+```
+
+### Update `update_chain_state_with()` for `Chain` struct to calculate Chain.value_pool
+
+- Location: `zebra-state/src/service/non_finalized_state/chain.rs`
+
+```rust
+fn update_chain_state_with(&mut self, prepared: &PreparedBlock) {        
+    ...
+    self.value_pool += prepared.block_value_balance;
+    ...
+}
+```
+
+### Check pool balance consensus rules
+
+#### Check that no pool is negative
+
+- Do the check in `commit_block` and `commit_new_chain`, located in `zebra-state/src/service/non_finalized_state.rs`. Both methods will need new a new argument with the value pool saved in the disk.
+- Pass this value to a new method `check_value_pools()` that will make the consensus rules check.
+
+```rust
+pub fn commit_block(
+    &mut self,
+    prepared: PreparedBlock,
+    finalized_tip_history_tree: &HistoryTree,
+    finalized_value_pool: ValueBalance<NonNegative>,
+) -> Result<(), ValidateContextError> {
+    ..
+    check_value_pools(finalized_value_pools, prepared.value_balance)?;
+    ..
+}
+
+pub fn commit_new_chain(
+    &mut self,
+    prepared: PreparedBlock,
+    finalized_tip_history_tree: HistoryTree,
+    finalized_value_pool: ValueBalance<NonNegative>,
+) -> Result<(), ValidateContextError> {
+    ..
+    check_value_pools(finalized_value_pools, prepared.value_balance)?;
```

Comment:
I have something wrong with the chain value pool and the block value pool. I think we need to pass the chain one here but not sure and i also think we dont have it at this stage.

Reply:
@oxarbitrage can you add this state change to the design?

@conradoplg can help if you have any questions about how it works.

Reply:
- add new parameter to the method `commit_new_chain()` which haves the balance of the finalized tip.
- pass the new argument to the `Chain` in:

`let mut chain = Chain::new(finalized_tip_history_tree, finalized_tip_value_balance);`

Reply:
I added what we discussed today for this in https://github.com/ZcashFoundation/zebra/pull/2430/commits/f814ca9f48a09b470667d8debfbfdfee96086922